### DISCUSSION TITLE
Setting nested state make the other states change (their values didn't change)

### DISCUSSION THREAD

**Discussion by khuongtp** [pmndrs/zustand#2506] [Bug report]
## Link to reproduction

https://codesandbox.io/p/sandbox/amazing-rhodes-6yp8c5?file=%2Fsrc%2FApp.tsx

Please view the console, you can see that the components that watch `count` re-render even though `count` is the same. Refresh the page in code sandbox make it work as expected but on my real machine it doesn't.
Here is my console log no matter how many times I reload the page
<img width="159" alt="image" src="https://github.com/pmndrs/zustand/assets/62872729/4e158798-df22-4797-9a89-2df692c6cc62">
(I'm using vite btw)
I did some research and found out that this piece of code was the problem, look at `count` in the useLayoutEffect dependencies, but I still think it's a bug cause the watched state literally didn't change
```tsx
const count = useTestStore((state) => {
  return state.count;
});
const setNest1 = useTestStore((state) => {
  return state.setNest1;
});

useLayoutEffect(() => {
  setNest1({ key: "nest2", value: { value1: 1 } });
}, [setNest1, count]);
```

## Check List

Please do not ask questions in issues.

- [ x] I've already opened a [discussion](https://github.com/pmndrs/zustand/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [ x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

Please check this if you're filing an issue regarding TypeScript.

- [ x] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

**Comment by dai-shi** [comment]
Isn't it because of useLayoutEffect?
Can you try wrapping console.log with useEffect and see if the result is expected?
```js
useEffect(() => {
  console.log('...');
});
```

**Reply by khuongtp** [reply]
Hi @dai-shi , I changed them all to useEffect but in TestComponent3 it still renders twice, Please hit reload a few times in code sandbox to see

**Reply by dai-shi** [reply]
How about changing useLayoutEffect to useEffect?

**Reply by khuongtp** [reply]
I already did https://codesandbox.io/p/sandbox/amazing-rhodes-6yp8c5?file=%2Fsrc%2FApp.tsx%3A74%2C14

**Reply by dbritto-dev** [reply]
@khuongtp this could happen since you are adding the set function to useEffect deps, I would use `useRef` to avoid to that

**Reply by khuongtp** [reply]
@dbritto-dev you mean like this 
```tsx
const setNest1 = useRef(useTestStore.getState().setNest1);

useEffect(() => {
  if (count === undefined) {
    return;
  }
  setNest1.current({ key: "nest2", value: { value1: 1 } });
}, [count]);
```
It still doesn't work

**Reply by dbritto-dev** [reply]
@khuongtp I don't see any issue -> 
<img width="1351" alt="image" src="https://github.com/pmndrs/zustand/assets/1697714/5ae935af-541f-4788-a165-1939793861e0">

**Reply by khuongtp** [reply]
Why TestComponent3 gets logged twice when `count` only changes from undefined to 999, and in the code I already checked if `count` is undefined then don't log

**Reply by dbritto-dev** [reply]
@khuongtp that's expected because by default we do shallow comparison

**Reply by dbritto-dev** [reply]
@khuongtp something to keep in mind if that you are calling a `setState` function twice at same time on mounting which means that at that point count is undefined in both calls that could lead to two triggers/renders

**Reply by dbritto-dev** [reply]
so I would say that's the current behavior is right there's no issue here

**Reply by dbritto-dev** [reply]
btw, here you go -> https://codesandbox.io/p/sandbox/amazing-rhodes-forked-xzrfj2
